### PR TITLE
Templates management

### PR DESCRIPTION
**Pull request by BaptisteViry**
### Description of your PR

Templates management in front
- [x] list templates
- [x] view template
- [x] new template
- [x] edit template

### Pre-review TODO

- [x] PR is rebased on top of `origin/master` 
- [ ] PR provide new tests for new behaviors
- [x] PR doesn't introduce commented out code
- [ ] PR has been tested locally on the main use cases
- [ ] Documentation have been updated if necessary

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      {
+        "name": "ng e2e",
+        "type": "node",
+        "request": "launch",
+        "program": "${workspaceFolder}/node_modules/protractor/bin/protractor",
+        "protocol": "inspector",
+        "args": ["${workspaceFolder}/protractor.conf.js"]
+      }
+    ]
+  }
```
Comment (Embraser01) [inline .vscode/launch.json]: Remove this file

File: client/src/app/app.routing.ts [hunk 2]
```diff
@@ -94,10 +95,13 @@ const routes: Routes = [
...
-        data: generateData('template:read') },
-    // { path: 'templates/new', component: TemplateNewComponent },
+        data: generateData(['template:read']) },
+    { path: 'templates/new', component: TemplateNewComponent, canActivate: [NgxPermissionsGuard],
+        data: generateData(['template:write']) },
     { path: 'templates/:id', component: TemplateViewComponent, canActivate: [NgxPermissionsGuard],
-        data: generateData('template:read') },
+        data: generateData(['template:read']) },
+    { path: 'templates/:id/edit', component: TemplateEditComponent, canActivate: [NgxPermissionsGuard],
+        data: generateData(['template:write']) },
```
Comment (Embraser01) [inline client/src/app/app.routing.ts]: We can do it without an array now no ?

File: client/src/app/templates/edit/template-edit.component.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        });
+        this.templateService.update(template).subscribe(() => {
+            this.toasterService.showToaster('Template modifié');
+            this.router.navigate(['/templates']);
+        });
+    }
+
+    prepareService(controls): {nbMax: Number,
+        startAt: {day: Number, hours: Number, minutes: Number},
+        endAt: {day: Number, hours: Number, minutes: Number}} {
```
Comment (Embraser01) [inline client/src/app/templates/edit/template-edit.component.ts]: It's weird to have the object described here

File: client/src/app/templates/new/templates-new.component.ts [hunk 4]
```diff
@@ -1,84 +1,116 @@
...
-        { id: '5', value: 'Vendredi' },
-        { id: '6', value: 'Samedi' },
-        { id: '7', value: 'Dimanche' } ];
+        { id: '0', value: 'Lundi' },
+        { id: '1', value: 'Mardi' },
+        { id: '2', value: 'Mercredi' },
+        { id: '3', value: 'Jeudi' },
+        { id: '4', value: 'Vendredi' },
+        { id: '5', value: 'Samedi' },
+        { id: '6', value: 'Dimanche' } ];
```
Comment (Embraser01) [inline client/src/app/templates/new/templates-new.component.ts]: Are you sur about using 0 for Monday? I think that most of libs use 0 for Sunday?

File: client/src/app/templates/new/templates-new.component.ts [hunk 5]
```diff
@@ -1,84 +1,116 @@
...
+    getControls(): Array<AbstractControl> {
+        return (this.generalFormArray.get([1]) as FormArray).controls;
+    }
+
+    removeServiceForm(fromGroupId: Number): void {
+        this.servicesFormArray.removeAt(+fromGroupId);
     }
 
-    toNumber(date: Date, selectedDay): { day: Number, hours: Number, minutes: Number } {
+    toNumber(date: String, selectedDay): { day: Number, hours: Number, minutes: Number } {
```
Comment (Embraser01) [inline client/src/app/templates/new/templates-new.component.ts]: Same here for the return type, it's less weird here but I think we should use an interface at the start of the file instead

File: client/src/app/templates/new/templates-new.component.ts [hunk 6]
```diff
@@ -1,84 +1,116 @@
...
         return {
             day: selectedDay,
-            hours: date.getHours(),
-            minutes: date.getMinutes(),
+            hours: Number(date.split(':')[0]),
+            minutes: Number(date.split(':')[1]),
```
Comment (Embraser01) [inline client/src/app/templates/new/templates-new.component.ts]: I saw the use of `+something` to conert to number earlier. We should maybe use the same pattern everywhere, what do you think?

File: client/src/app/templates/new/templates-new.component.ts [hunk 7]
```diff
@@ -1,84 +1,116 @@
...
+        });
+        this.templateService.create(template).subscribe(() => {
+            this.toasterService.showToaster('Template créé');
+            this.router.navigate(['/templates']);
+        });
+    }
 
+    prepareService(controls): {nbMax: Number,
+        startAt: {day: Number, hours: Number, minutes: Number},
+        endAt: {day: Number, hours: Number, minutes: Number}} {
```
Comment (Embraser01) [inline client/src/app/templates/new/templates-new.component.ts]: Same. Interface?

File: .vscode/launch.json (hunk 1 above)
Comment (kistora) [reply .vscode/launch.json]: Sorry :)

File: client/src/app/app.routing.ts (hunk 2 above)
Comment (kistora) [reply client/src/app/app.routing.ts]: Yes !

File: client/src/app/templates/edit/template-edit.component.ts (hunk 3 above)
Comment (kistora) [reply client/src/app/templates/edit/template-edit.component.ts]: You advise to create a new model for TemplateUnit ?

File: client/src/app/templates/new/templates-new.component.ts (hunk 4 above)
Comment (kistora) [reply client/src/app/templates/new/templates-new.component.ts]: Hum...
Let's change that !

File: .vscode/launch.json (hunk 1 above)
Comment (Embraser01) [reply .vscode/launch.json]: Actually it's not really your fault, it's in the gitignore, can you remove exceptions here:

https://github.com/K-Fet/K-App/blob/2a147461744dce36b5d046cc1f0e328f3da7c0a6/.gitignore#L26-L30

File: client/src/app/templates/edit/template-edit.component.ts (hunk 3 above)
Comment (Embraser01) [reply client/src/app/templates/edit/template-edit.component.ts]: For this one yes we could

File: client/src/app/templates/new/templates-new.component.ts (hunk 4 above)
Comment (kistora) [reply client/src/app/templates/new/templates-new.component.ts]: image

So we don't change it :)

Comment (Embraser01) [reply client/src/app/templates/new/templates-new.component.ts]: So you should revert the changes 😸

File: .vscode/launch.json (hunk 1 above)
Comment (kistora) [reply .vscode/launch.json]: Implemented

File: client/src/app/templates/edit/template-edit.component.ts (hunk 3 above)
Comment (kistora) [reply client/src/app/templates/edit/template-edit.component.ts]: I've created an interface inside the file.

File: client/src/app/templates/new/templates-new.component.ts (hunk 4 above)
Comment (kistora) [reply client/src/app/templates/new/templates-new.component.ts]: Yes :) I've changed the BACK template model (get & set)

File: client/src/app/templates/new/templates-new.component.ts (hunk 5 above)
Comment (kistora) [reply client/src/app/templates/new/templates-new.component.ts]: I've created an interface again :)

File: client/src/app/templates/new/templates-new.component.ts (hunk 6 above)
Comment (kistora) [reply client/src/app/templates/new/templates-new.component.ts]: Yes

File: .gitignore [hunk 8]
```diff
@@ -207,3 +207,10 @@ fabric.properties
 
 # Ignore backups
 backups/
+
+# Ignore VS Code config file
+.vscode/*
+ !.vscode/settings.json
```
Comment (Embraser01) [inline .gitignore]: Sorry, that is not what I meant, It is already there and you must remove the lines starting with `!.vscode[...]`

File: server/app/models/template.js [hunk 9]
```diff
@@ -69,9 +69,9 @@ class TemplateUnit extends Model {
                 // eslint-disable-next-line
                 get() {
                     const startAt = this.getDataValue('startAt');
-
+                    const day = startAt.getDay() === 0 ? 7 : startAt.getDay();
```
Comment (Embraser01) [inline server/app/models/template.js]: Can be simplified to `const day = startAt.getDay() || 7`

File: server/app/models/template.js [hunk 10]
```diff
@@ -69,9 +69,9 @@ class TemplateUnit extends Model {
...
                     return {
-                        day: startAt.getDay(),
+                        day: day,
```
Comment (Embraser01) [inline server/app/models/template.js]: Convert to shorthand property

File: server/app/models/template.js [hunk 11]
```diff
@@ -95,9 +95,9 @@ class TemplateUnit extends Model {
                 // eslint-disable-next-line
                 get() {
                     const endAt = this.getDataValue('endAt');
-
+                    const day = endAt.getDay() === 0 ? 7 : endAt.getDay();
```
Comment (Embraser01) [inline server/app/models/template.js]: Same for shorthand & simplification

File: server/app/models/template.js (hunk 10 above)
Comment (kistora) [reply server/app/models/template.js]: Smth like that:
```
const day = endAt.getDay() || 7;
                    return {
                        day,
                        hours: endAt.getHours(),
                        minutes: endAt.getMinutes(),
                    };
```

File: server/app/models/template.js (hunk 9 above)
Comment (kistora) [reply server/app/models/template.js]: Oh wonderfull :)

**Pull request merged**